include/nbl/builtin/hlsl/workgroup/shared_scan.hlsl

```diff
@@ -0,0 +1,184 @@
+        // the first SubgroupSz^2 invocations will be processed by the first subgroup and so on.
+        // Consequently, those first SubgroupSz^2 invocations will store their results on SubgroupSz scratch slots 
+        // with halfSubgroupSz padding and the next level will follow the same + the previous as an `offset`.
+        const uint offset = (gl_LocalInvocationIndex >> glsl::gl_SubgroupSizeLog2()) & ~subgroupMask; // For subgroupSz = 64, first 4095 invocations get offset 0, 4096-8191 offset 64, then 128 etc.
+        const uint memBegin = offset;
+        uint nextLevelStoreIndex = memBegin + glsl::gl_SubgroupID();
+        scanStoreIndex = gl_LocalInvocationIndex;
+        
+        bool participate = gl_LocalInvocationIndex <= lastInvocationInLevel;
+        while(lastInvocationInLevel >= glsl::gl_SubgroupSize() * glsl::gl_SubgroupSize())
+        {
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            if(participate)
+            {
+                if (any(bool2(gl_LocalInvocationIndex == lastInvocationInLevel, isLastSubgroupInvocation)))
+                { // only invocations that have the final value of the subgroupOp (inclusive scan) store their results
+                    sharedAccessor.main.set(nextLevelStoreIndex, scan);
+                }
+            }
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
+            if(participate)
+            {
+                const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
+                scan = subgroupOp(prevLevelScan);
+                if(isScan)
+                {
+                    sharedAccessor.main.set(scanStoreIndex, scan);
+                }
+            }
+            if(isScan)
+            {
+                scanStoreIndex += lastInvocationInLevel + 1u;
+            }
+        }
+        if(lastInvocationInLevel >= glsl::gl_SubgroupSize())
+        {
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            if(participate)
+            {
+                if(any(bool2(gl_LocalInvocationIndex == lastInvocationInLevel, isLastSubgroupInvocation))) {
+                    sharedAccessor.main.set(nextLevelStoreIndex, scan);
+                }
+            }
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
+            if(participate)
+            {
+                const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
+                scan = subgroupOp(prevLevelScan);
+                if(isScan) {
+                    sharedAccessor.main.set(scanStoreIndex, scan);
+                }
+            }
+        }
+        return scan;
```

Comment:
add a comment that only invocations of `gl_LocalInvocationIndex<gl_SubgroupCount` will return correct values, rest will be "garbage"

Reply:
actually there's no use in returning this `scan`, just make the operator VOID

Reply:
but turn `scan` into a member (so we can pass it to TAIL or do broadcast that way), put the same comment about validity there